File: wrappers/java/foundation/src/test/java/com/virgilsecurity/crypto/foundation/Ed25519Test.java
/*
* Copyright (C) 2015-2019 Virgil Security, Inc.
*
* All rights reserved.
*
* Redistribution and use in source and binary forms, with or without
* modification, are permitted provided that the following conditions are
* met:
*
* (1) Redistributions of source code must retain the above copyright
* notice, this list of conditions and the following disclaimer.
*
* (2) Redistributions in binary form must reproduce the above copyright
* notice, this list of conditions and the following disclaimer in
* the documentation and/or other materials provided with the
* distribution.
*
* (3) Neither the name of the copyright holder nor the names of its
* contributors may be used to endorse or promote products derived from
* this software without specific prior written permission.
*
* THIS SOFTWARE IS PROVIDED BY THE AUTHOR ''AS IS'' AND ANY EXPRESS OR
* IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
* WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
* DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
* INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
* (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
* SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
* HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
* STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
* IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
* POSSIBILITY OF SUCH DAMAGE.
*
* Lead Maintainer: Virgil Security Inc. <support@virgilsecurity.com>
*/

package com.virgilsecurity.crypto.foundation;

import static org.junit.Assert.assertArrayEquals;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

import java.util.Random;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;

public class Ed25519Test extends SampleBasedTest {

	private Ed25519 ed;
	private KeyProvider keyProvider;

	@Before
	public void init() {
		this.ed = new Ed25519();
		this.ed.setupDefaults();

		this.keyProvider = new KeyProvider();
		this.keyProvider.setupDefaults();
	}

	@After
	public void tearDown() {
		this.ed.close();
		this.keyProvider.close();
	}

	@Test
	public void generateKey() {
		PrivateKey privateKey = this.ed.generateKey();
		assertNotNull(privateKey);
		assertEquals(AlgId.ED25519, privateKey.algId());
	}

	@Test
	public void algId() {
		assertEquals(AlgId.ED25519, this.ed.algId());
	}

	@Test
	public void canSign() {
		PrivateKey privateKey = this.ed.generateKey();
		assertTrue(this.ed.canSign(privateKey));
	}

	@Test
	public void canSign_wrongKey() {
		try (Rsa rsa = new Rsa()) {
			rsa.setupDefaults();
			PrivateKey privateKey = rsa.generateKey(2048);
			assertFalse(this.ed.canSign(privateKey));
		}
	}

	@Test
	public void getCanExportPrivateKey() {
		assertTrue(this.ed.getCanExportPrivateKey());
	}

	@Test
	public void getCanImportPrivateKey() {
		assertTrue(this.ed.getCanImportPrivateKey());
	}

	@Test
	public void export_import_PrivateKey() {
		PrivateKey privateKey = this.ed.generateKey();

		// Export private key
		RawPrivateKey rawPrivateKey = this.ed.exportPrivateKey(privateKey);
		assertNotNull(rawPrivateKey);

		byte[] exportedKeyData = rawPrivateKey.data();
		assertNotNull(exportedKeyData);

		try (RawPrivateKey importedPrivateKey = (RawPrivateKey) this.ed.importPrivateKey(rawPrivateKey)) {
			RawPrivateKey rawPrivateKey2 = this.ed.exportPrivateKey(importedPrivateKey);
			assertNotNull(rawPrivateKey2);

			byte[] exportedKeyData2 = rawPrivateKey2.data();
			assertNotNull(exportedKeyData2);
			assertArrayEquals(exportedKeyData, exportedKeyData2);
		}
	}

	@Test
	public void export_import_PublicKey() {
		PrivateKey privateKey = this.ed.generateKey();
		PublicKey publicKey = privateKey.extractPublicKey();

		// Export private key
		RawPublicKey rawPublicKey = this.ed.exportPublicKey(publicKey);
		assertNotNull(rawPublicKey);

		byte[] exportedKeyData = rawPublicKey.data();
		assertNotNull(exportedKeyData);

		try (RawPublicKey importedPublicKey = (RawPublicKey) this.ed.importPublicKey(rawPublicKey)) {
			RawPublicKey rawPublicKey2 = this.ed.exportPublicKey(importedPublicKey);
			assertNotNull(rawPublicKey2);

			byte[] exportedKeyData2 = rawPublicKey2.data();
			assertNotNull(exportedKeyData2);
			assertArrayEquals(exportedKeyData, exportedKeyData2);
		}
	}

	@Test
	public void encrypt_decrypt() {
		byte[] data = new byte[100];
		new Random().nextBytes(data);

		try (RawPrivateKey privateKey = (RawPrivateKey) this.ed.generateKey();
				RawPublicKey publicKey = (RawPublicKey) privateKey.extractPublicKey()) {

			assertTrue(this.ed.canEncrypt(publicKey, data.length));

			byte[] encryptedData = this.ed.encrypt(publicKey, data);
			assertNotNull(encryptedData);

			assertTrue(this.ed.canDecrypt(privateKey, encryptedData.length));

			byte[] decryptedData = this.ed.decrypt(privateKey, encryptedData);
			assertNotNull(decryptedData);

			assertArrayEquals(data, decryptedData);
		}
	}

	@Test
	public void decrypt() {
		byte[] privateKeyData = getBytes("ed25519.private_key");
		byte[] encryptedData = getBytes("ed25519.encrypted_data");
		byte[] expectedDecryptedData = getBytes("short_data");

		try (RawPrivateKey privateKey = (RawPrivateKey) keyProvider.importPrivateKey(privateKeyData)) {

			byte[] decryptedData = this.ed.decrypt(privateKey, encryptedData);
			assertNotNull(decryptedData);
			assertArrayEquals(expectedDecryptedData, decryptedData);
		}
	}

	@Test
	public void sign() {
		byte[] data = getBytes("data");
		byte[] privateKeyData = getBytes("ed25519.private_key");

		try (RawPrivateKey privateKey = (RawPrivateKey) keyProvider.importPrivateKey(privateKeyData);
				RawPublicKey publicKey = (RawPublicKey) privateKey.extractPublicKey()) {

			assertTrue(this.ed.canSign(privateKey));

			byte[] signature = this.ed.signHash(privateKey, AlgId.SHA512, data);
			assertNotNull(signature);
			assertEquals(this.ed.signatureLen(privateKey), signature.length);

			assertTrue(this.ed.verifyHash(publicKey, AlgId.SHA512, data, signature));
		}
	}

	@Test
	public void verifyHash() {
		byte[] data = getBytes("data");
		byte[] publicKeyData = getBytes("ed25519.public_key");
		byte[] signature = getBytes("ed25519.signature");

		try (RawPublicKey publicKey = (RawPublicKey) keyProvider.importPublicKey(publicKeyData)) {
			assertTrue(this.ed.canVerify(publicKey));

			assertTrue(this.ed.verifyHash(publicKey, AlgId.SHA512, data, signature));
		}
	}

	@Test
	public void verifyHash_wrongSignature() {
		byte[] data = getBytes("data");
		byte[] publicKeyData = getBytes("ed25519.public_key");
		byte[] signature = getBytes("ed25519.wrong_signature");

		try (RawPublicKey publicKey = (RawPublicKey) keyProvider.importPublicKey(publicKeyData)) {
			assertFalse(this.ed.verifyHash(publicKey, AlgId.SHA512, data, signature));
		}
	}

}

File: wrappers/go/foundation/ed25519_test.go
//  Copyright (C) 2015-2020 Virgil Security, Inc.
//
//  All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are
//  met:
//
//      (1) Redistributions of source code must retain the above copyright
//      notice, this list of conditions and the following disclaimer.
//
//      (2) Redistributions in binary form must reproduce the above copyright
//      notice, this list of conditions and the following disclaimer in
//      the documentation and/or other materials provided with the
//      distribution.
//
//      (3) Neither the name of the copyright holder nor the names of its
//      contributors may be used to endorse or promote products derived from
//      this software without specific prior written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE AUTHOR ''AS IS'' AND ANY EXPRESS OR
//  IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
//  WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
//  DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
//  INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
//  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
//  HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
//  STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
//  IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
//  POSSIBILITY OF SUCH DAMAGE.
//
//  Lead Maintainer: Virgil Security Inc. <support@virgilsecurity.com>

package foundation

import (
	b64 "encoding/base64"
	"math/rand"
	"testing"

	"github.com/stretchr/testify/require"
)

func TestNewEd25519(t *testing.T) {
	ed := NewEd25519()

	require.NotNil(t, ed)
}

func TestEd25519_GenerateKey(t *testing.T) {
	ed := newEd25519()

	privateKey, err := ed.GenerateKey()
	require.Nil(t, err)
	require.NotNil(t, privateKey)

	edKey, ok := privateKey.(Key)
	require.True(t, ok)
	require.Equal(t, AlgIdEd25519, edKey.AlgId())
}

func TestEd25519_CanSign(t *testing.T) {
	ed := newEd25519()
	privateKey, err := ed.GenerateKey()
	require.Nil(t, err)
	require.True(t, ed.CanSign(privateKey))
}

func TestEd25519_CanSign_WrongKey(t *testing.T) {
	ed := newEd25519()
	rsa := NewRsa()
	err := rsa.SetupDefaults()
	require.Nil(t, err)

	privateKey, err := rsa.GenerateKey(2048)
	require.Nil(t, err)
	require.False(t, ed.CanSign(privateKey))
}

func TestEd25519_GetCanExportPrivateKey(t *testing.T) {
	ed := newEd25519()

	require.True(t, ed.GetCanExportPrivateKey())
}

func TestEd25519_GetCanImportPrivateKey(t *testing.T) {
	ed := newEd25519()

	require.True(t, ed.GetCanImportPrivateKey())
}

func TestEd25519_ExportPrivateKey(t *testing.T) {
	ed := newEd25519()
	privateKey, err := ed.GenerateKey()
	require.Nil(t, err)

	// Export private key
	rawPrivateKey, err := ed.ExportPrivateKey(privateKey)
	require.Nil(t, err)
	require.NotNil(t, rawPrivateKey)

	exportedKeyData := rawPrivateKey.Data()
	require.NotNil(t, exportedKeyData)

	importedPrivateKey, err := ed.ImportPrivateKey(rawPrivateKey)
	require.Nil(t, err)

	rawPrivateKey2, err := ed.ExportPrivateKey(importedPrivateKey)
	require.Nil(t, err)
	require.NotNil(t, rawPrivateKey2)

	exportedKeyData2 := rawPrivateKey2.Data()

	require.NotNil(t, exportedKeyData2)
	require.Equal(t, exportedKeyData, exportedKeyData2)
}

func TestEd25519_ExportPublicKey(t *testing.T) {
	ed := newEd25519()
	privateKey, err := ed.GenerateKey()
	require.Nil(t, err)

	publicKey, err := privateKey.ExtractPublicKey()
	require.Nil(t, err)

	// Export public key
	rawPublicKey, err := ed.ExportPublicKey(publicKey)
	require.Nil(t, err)
	require.NotNil(t, rawPublicKey)

	exportedKeyData := rawPublicKey.Data()
	require.NotNil(t, exportedKeyData)

	importedPublicKey, err := ed.ImportPublicKey(rawPublicKey)
	require.Nil(t, err)

	rawPublicKey2, err := ed.ExportPublicKey(importedPublicKey)
	require.Nil(t, err)
	require.NotNil(t, rawPublicKey2)

	exportedKeyData2 := rawPublicKey2.Data()

	require.NotNil(t, exportedKeyData2)
	require.Equal(t, exportedKeyData, exportedKeyData2)
}

func TestEd25519_Encrypt(t *testing.T) {
	data := make([]byte, 100)
	rand.Read(data)

	ed := newEd25519()
	privateKey, err := ed.GenerateKey()
	require.Nil(t, err)

	publicKey, err := privateKey.ExtractPublicKey()
	require.Nil(t, err)

	require.True(t, ed.CanEncrypt(publicKey, uint(len(data))))

	encryptedData, err := ed.Encrypt(publicKey, data)
	require.Nil(t, err)
	require.NotNil(t, encryptedData)

	require.True(t, ed.CanDecrypt(privateKey, uint(len(encryptedData))))

	decryptedData, err := ed.Decrypt(privateKey, encryptedData)
	require.Nil(t, err)
	require.NotNil(t, decryptedData)

	require.Equal(t, data, decryptedData)
}

func TestEd25519_Decrypt(t *testing.T) {
	privateKeyData, _ := b64.StdEncoding.DecodeString(TEST_ED25519_PRIVATE_KEY)
	encryptedData, _ := b64.StdEncoding.DecodeString(TEST_ED25519_ENCRYPTED_DATA)
	expectedDecryptedData, _ := b64.StdEncoding.DecodeString(TEST_SHORT_DATA)
	ed := newEd25519()
	keyProvider := newKeyProvider()

	privateKey, err := keyProvider.ImportPrivateKey(privateKeyData)
	require.Nil(t, err)

	decryptedData, err := ed.Decrypt(privateKey, encryptedData)
	require.Nil(t, err)
	require.NotNil(t, decryptedData)
	require.Equal(t, expectedDecryptedData, decryptedData)
}

func TestEd25519_SignHash(t *testing.T) {
	data, _ := b64.StdEncoding.DecodeString(TEST_DATA)
	privateKeyData, _ := b64.StdEncoding.DecodeString(TEST_ED25519_PRIVATE_KEY)
	ed := newEd25519()
	keyProvider := newKeyProvider()

	privateKey, err := keyProvider.ImportPrivateKey(privateKeyData)
	require.Nil(t, err)

	publicKey, err := privateKey.ExtractPublicKey()
	require.Nil(t, err)
	require.True(t, ed.CanSign(privateKey))

	signature, err := ed.SignHash(privateKey, AlgIdSha512, data)
	require.Nil(t, err)
	require.NotNil(t, signature)
	require.Equal(t, ed.SignatureLen(privateKey), uint(len(signature)))

	require.True(t, ed.VerifyHash(publicKey, AlgIdSha512, data, signature))
}

func TestEd25519_VerifyHash(t *testing.T) {
	data, _ := b64.StdEncoding.DecodeString(TEST_DATA)
	publicKeyData, _ := b64.StdEncoding.DecodeString(TEST_ED25519_PUBLIC_KEY)
	signature, _ := b64.StdEncoding.DecodeString(TEST_ED25519_SIGNATURE)
	ed := newEd25519()
	keyProvider := newKeyProvider()

	publicKey, err := keyProvider.ImportPublicKey(publicKeyData)
	require.Nil(t, err)
	require.True(t, ed.CanVerify(publicKey))
	require.True(t, ed.VerifyHash(publicKey, AlgIdSha512, data, signature))
}

func TestEd25519_VerifyHash_WrongHash(t *testing.T) {
	data, _ := b64.StdEncoding.DecodeString(TEST_DATA)
	publicKeyData, _ := b64.StdEncoding.DecodeString(TEST_ED25519_PUBLIC_KEY)
	signature, _ := b64.StdEncoding.DecodeString(TEST_ED25519_WRONG_SIGNATURE)
	ed := newEd25519()
	keyProvider := newKeyProvider()

	publicKey, err := keyProvider.ImportPublicKey(publicKeyData)
	require.Nil(t, err)
	require.False(t, ed.VerifyHash(publicKey, AlgIdSha512, data, signature))

}
func newEd25519() *Ed25519 {
	ed := NewEd25519()
	_ = ed.SetupDefaults()

	return ed
}
